src/Tools/SarifValidator/ResultLogBuilder.cs

```diff
@@ -0,0 +1,258 @@
+                null);          // The version of this tool has no prerelease info.
+        }
+
+        internal IEnumerable<string> BuildLog(List<JsonError> errors)
+        {
+            foreach (JsonError error in errors)
+            {
+                LogError(error);
+            }
+
+            return _messages;
+        }
+
+        private NewLineIndex InstanceFileIndex
+        {
+            get
+            {
+                if (_instanceFileIndex == null)
+                {
+                    _instanceFileIndex = new NewLineIndex(_fileSystem.ReadAllText(_options.InstanceFilePath));
+                }
+
+                return _instanceFileIndex;
+            }
+        }
+
+        private NewLineIndex SchemaFileIndex
+        {
+            get
+            {
+                if (_schemaFileIndex == null)
+                {
+                    _schemaFileIndex = new NewLineIndex(_fileSystem.ReadAllText(_options.SchemaFilePath));
+                }
+
+                return _schemaFileIndex;
+            }
+        }
+
+        private void LogError(JsonError error)
+        {
+            IRuleDescriptor rule = GetRuleDescriptorForError(error);
+            Result result = MakeResultFromError(error);
+
+            _logger.Log(rule, result);
+            _messages.Add(result.FormatForVisualStudio(rule));
+        }
+
+        private static IRuleDescriptor GetRuleDescriptorForError(JsonError error)
+        {
+            switch (error.Kind)
+            {
+                case JsonErrorKind.Syntax:
+                    return JsonSyntaxErrorRule;
+
+                case JsonErrorKind.Validation:
+                    return JsonSchemaValidationErrorRule;
+
+                default:
+                    return UnknownErrorRule;
+            }
+        }
+
+        private Result MakeResultFromError(JsonError error)
+        {
+            var result = new Result();
+
+            string analysisTargetFilePath;
+            NewLineIndex index = null;
+
+            switch (error.Location)
+            {
+                case JsonErrorLocation.InstanceDocument:
+                    analysisTargetFilePath = _options.InstanceFilePath;
+                    index = InstanceFileIndex;
+                    break;
+
+                case JsonErrorLocation.Schema:
+                    analysisTargetFilePath = _options.SchemaFilePath;
+                    index = SchemaFileIndex;
+                    break;
+
+                default:
+                    analysisTargetFilePath = "unknown_file";
+                    break;
+            }
+
+            // Subtle: We make a Uri object out of the file path. The AbsoluteUri property
+            // of the resulting Uri is a string representation of the resulting "file://"
+            // Uri. We then convert that string into another Uri object.
+            Uri analysisTargetUri = new Uri(new Uri(analysisTargetFilePath).AbsoluteUri);
```

Comment:
there's a helper in the SDK that already does this. it appears likely this subtlety will be a persistent source of irritation for users. how do we address?